### PR TITLE
test(#100): Storybook interaction (play function) tests and the test-storybook gate

### LINKED ISSUE #100: NFR(Testing/Verification): zero Storybook interaction (play) tests — component behavior is never exercised where it is demonstrated
## Problem

The library's 38 stories across 25 components are the public demonstration surface, but none of them verify behavior. Evidence:

- `grep -rn 'play:' src --include='*.stories.*'` → zero play functions in any story.
- `package.json:27-32` declares only `@storybook/addon-docs`, `addon-links`, `addon-onboarding`, `addon-webpack5-compiler-swc`; there is no `@storybook/test`, no `@storybook/test-runner`, and `.storybook/main.ts` registers no interaction tooling.
- The e2e smoke suite (`tests/e2e/stories.smoke.spec.ts:14-32`) only asserts each story **mounts** (`childElementCount > 0`, no pageerror). Visual tests assert static pixels. Neither clicks a button, types into `UiInput`, toggles `UiCheckbox`, submits `UiForm`, or opens `UiTooltip` *as rendered in the story with its story args*.

Unit/integration tests (jsdom) cover component logic at 100%, but jsdom does not exercise real browser event handling, Emotion-styled focus/hover state machines, or MUI portal behavior (tooltip/select popovers). For a component library, interaction tests in the real Storybook build are the standard way to prove "the component works as demoed" — and they double as living documentation in the Interactions panel.

## Quality attributes blocked

- **testability** (currently 4/5): interactive behavior inside stories is untestable today.
- **correctness** (4/5): behavior verified only in jsdom, not in the shipped Storybook browser context.
- **demonstrability** (4/5): stories show appearance but never demonstrate working behavior.

## Proposed solution

1. Add `@storybook/test` and write `play` functions for every interactive component's primary story — at minimum: `UiButton` (click fires `onClick`), `UiInput`/`UiTextFieldForm` (type + value + error state), `UiCheckbox` (toggle), `UiForm` (fill + submit + validation error path), `UiTooltip` (hover/focus opens), `UiLink`, `UiBackToMain`, `UiSelect`/toolbar controls as applicable. Use `within(canvasElement)` + `userEvent` + `expect` from `@storybook/test`.
2. Run them in CI against the real built Storybook via `@storybook/test-runner` (Playwright-based — reuse the existing `Dockerfile.playwright` + `run-storybook-playwright` pattern in `Makefile:44-55`) as a new `make test-storybook` target, wired into a workflow and into `make verify`.
3. Fail-if-absent from day one (consistent with the silent-skip hardening issue): the test-runner invocation must fail if it discovers zero play functions; add a small manifest check (like `tests/visual/visual.spec.ts:52-71`) listing which stories are expected to carry interactions so coverage can only grow.
4. Keep scope: functional interactions only. Axe/keyboard/focus-order a11y assertions belong to #66 and must not be duplicated here (the test-runner's a11y hook can be enabled under #66 later).

## Acceptance criteria

- [ ] `@storybook/test` and `@storybook/test-runner` (or portable-stories equivalent) added; versions caret-pinned per `scripts/ci/check-dependency-ranges.ts` policy.
- [ ] ≥ 8 interactive components have `play` functions asserting behavior via `userEvent` + `expect`.
- [ ] `make test-storybook` runs the interaction suite headlessly in the Docker Playwright image and exits non-zero on any failure or on zero-tests-discovered.
- [ ] New GitHub Actions workflow (or job) runs `make test-storybook` on PRs to main with no conditional skip path.
- [ ] Interactions are visible in the Storybook Interactions panel for the covered stories.
- [ ] `tests/bats/make-target-coverage.tsv` / Bats contract updated for the new target.

## References

- #66 (a11y assertions in stories — explicitly out of scope here), #70 umbrella audit, #13-#30 component stories (each gains a behavioral proof)
- Files: `src/components/*/*.stories.tsx`, `.storybook/main.ts`, `package.json`, `Makefile:44-55`, `tests/e2e/stories.smoke.spec.ts`

### PR DESCRIPTION
**Pull request by RudoiDmytro**
Closes #100.

## What

The library's stories were the public demonstration surface but verified nothing:
`grep -rn 'play:' src --include='*.stories.*'` returned zero hits, and the e2e
smoke suite only asserted that each story *mounts*. This adds behavioural proof
where the behaviour is demonstrated, and a gate that cannot silently skip.

### 1. Play functions for 13 interactive components

A dedicated story per component, tagged `['interaction', '!autodocs']`, driving
the component with `userEvent` and asserting with `expect`:

| Component | Story | Asserts |
| --- | --- | --- |
| `UiButton` | `ClickInvokesHandler` | click dispatches `onClick` once |
| `UiInput` | `TypingUpdatesValue` | typed value lands; one `onChange` per keystroke |
| `UiCheckbox` | `ClickTogglesCheckedState` | click flips `checked` and reports it |
| `UiForm` | `ValidationBlocksEmptySubmit` | empty submit shows the field error and does not call `onSubmit`; a valid value does |
| `UiTextFieldForm` | `ValidationSurfacesFieldError` | validation message appears on submit and clears when the value becomes valid |
| `UiTooltip` | `ClickOpensTooltip` | closed at rest, opens on click, `aria-expanded` follows |
| `UiLink` | `KeyboardFocusExposesNewTabHint` | keyboard reachable; new-tab hint in the accessible name; `rel="noopener noreferrer"` |
| `UiBackToMain` | `KeyboardFocusReachesBackLink` | keyboard reachable link pointing at `to` |
| `UiSearchInput` | `SuggestionPickFillsField` | typing filters suggestions; picking one fills the field |
| `UiSelectWithSearch` | `SearchNarrowsAndSelectsOption` | search narrows the listbox; the pick becomes the value |
| `UiMultiSelect` | `PickingOptionsAddsChips` | two picks become two chips |
| `UiRadioGroup` | `ChoosingOptionMovesSelection` | choosing checks one and unchecks the previous |
| `UiCalendarMultiSelect` | `DayToggleUpdatesSelection` | a day toggles on and back off via `aria-selected` |

They are **new** exports, never `play` bolted onto an existing story: Storybook
autoplays `play` as soon as a canvas renders, which would destabilise the pinned
visual baseline of the story it demos. The four always-controlled field controls
get a stateful render wrapper so typing/selection actually sticks, and every
asserted handler is an explicit `fn()` spy (an implicit action arg throws when a
play function invokes it).

### 2. `make test-storybook` — fail-closed

Boots the shared Storybook service and runs `scripts/ci/run-storybook-interactions.ts`
in the Playwright image, reusing the existing `run-storybook-playwright` helper
(extended with a `PLAYWRIGHT_RUN_CMD` hook; the default path is byte-identical).

`@storybook/test-runner --includeTags interaction` rewrites every story file
without a matching story into a **skipped no-op suite**, so Jest alone exits `0`
when the tag disappears library-wide. The script therefore brackets the run:

1. **before** — the live Storybook index must expose exactly the stories
   registered in `tests/storybook/interaction-stories.json`, covering at least
   8 components;
2. **run** — `--includeTags interaction`, chromium, JUnit reporter;
3. **after** — the JUnit report must contain one *passing* `play-test` per
   registered story.

Verified failure modes (all exit `1`): empty registry, registry/index drift,
component count below the floor, any failing or skipped play test.

### 3. Coverage can only grow

- `tests/unit/storybook-interaction-coverage.test.ts` scans the story sources
  with the TypeScript AST (no module evaluation, so Stryker never re-runs it per
  mutant): a `play` function that is not tagged `interaction` **and** registered
  fails the unit suite.
- `tests/visual/visual.spec.ts` keeps its manifest-completeness gate untouched and
  adds a reciprocal one: the only stories the screenshot loop skips are the
  registered interaction stories, and every registered id is a real story.
  Interaction stories are pixel-exempt because their canvas mutates itself
  asynchronously — a screenshot races the `play` phase — and they keep their e2e
  smoke coverage on all three engines.

### 4. Contracts and docs

`tests/bats/make-target-coverage.tsv` row, two new `makefile_targets.bats` tests
(the target runs the gate, and the shared helper still defaults to Playwright for
`test-e2e`/`test-visual`), `test-storybook` added to `GATING_TARGETS`, plus
README / agents.md / CONTRIBUTING.md / `tests/storybook/README.md`.

## Incidental unblock

`main`'s Docker builds are currently broken: the pinned `nodejs=22.23.0-r0` and
`python3=3.12.13-r0` were dropped from the Alpine v3.22 index, so every
Docker-backed job fails at `apk add`. Bumped to the patch levels the index now
carries (verified with `apk add -s`). Separately, the Storybook node API eagerly
creates `node_modules/.cache/storybook`, which `pwuser` could not write — that
directory is now pre-created and chowned in `Dockerfile.playwright` (node_modules
itself stays root-owned).

## Acceptance criteria

- [x] Runner added, caret-pinned (`@storybook/test-runner@^0.24.4`). **Note:**
      `@storybook/test` tops out at 8.6.x and peers on `storybook@^8`; Storybook 10's
      equivalent is the `storybook/test` subpath of the already-installed
      `storybook@10.4.3`, which is what the play functions import.
- [x] 13 interactive components have `play` functions using `userEvent` + `expect` (≥ 8).
- [x] `make test-storybook` runs headlessly in the Docker Playwright image and exits
      non-zero on any failure or on zero tests discovered.
- [x] New workflow runs it on PRs to `main` with no conditional skip path.
- [x] Interactions appear in the Storybook Interactions panel (built into Storybook 10 core).
- [x] `tests/bats/make-target-coverage.tsv` and the Bats contract updated.

Axe / keyboard / focus-order a11y assertions stay out of scope per the issue (#66).

## Verification

Run locally against Docker on this branch:

| Check | Result |
| --- | --- |
| `make test-storybook` | 13 passed |
| `make test-e2e` | 216 passed (3 engines) |
| `make test-visual` | 71 passed, 130 skipped |
| `make lint-metrics` | all hard checks pass |
| `make lint-deps` | 0 errors |
| Bats suite | 271 passed |
| Jest unit | 624 passed, 100% coverage |
| Jest integration | 70 passed, 100% coverage |
| tsc / ESLint / Prettier / markdownlint | clean |

---
## Summary by cubic
Adds Storybook interaction tests (play functions) for 13 components and a fail-closed `make test-storybook` gate so browser-proven behavior cannot silently disappear. Previously stories only mounted; now each registered interaction story must run and pass in CI, and the gate fails on registry/index drift, duplicate entries, coverage below the floor, or any missing/failed/skipped play test.

- Adds `interaction`-tagged `play` stories for 13 components using `userEvent`/`expect` from `storybook/test`; controlled inputs use stateful wrappers and asserted handlers use explicit `fn()` spies. Stories are separate exports to avoid autoplay breaking visual baselines; a shared body in `.storybook/field-typeahead-interactions.ts` drives typeahead fields.
- Introduces `@storybook/test-runner` and `scripts/ci/run-storybook-interactions.ts`: reconciles the live index with `tests/storybook/interaction-stories.json`, enforces a ≥8-component floor, runs chromium only, then requires one passing play test per entry from JUnit and reports the distinct passed count. Hardening: manifest path is contained to the repo root with lexical and `realpathSync` checks, runner spawned via `process.execPath`, trailing slashes trimmed, output sanitized, duplicate rows rejected in both id and title+exportName key spaces, JUnit names XML‑decoded, explicit `localeCompare`/`String#endsWith` comparisons, and the count deduplicates repeated report rows.
- Drift guards: a static scan of CSF stories across all flavors (`*.stories.{js,jsx,mjs,ts,tsx}`) requires `play` stories to be tagged `interaction` and registered; it reads meta titles from inline defaults, resolves `export default meta` and `export { meta as default }` back to the declaration, and sees through `as`/`satisfies`. Stories are copied into the Stryker sandbox (still excluded from `mutate`). Dependency-cruiser and Jest/tsconfig exclusions now mirror the same multi-flavor patterns (including `.mjs`).
- Visual suite exempts registered interaction stories and asserts the exemption matches the registry; all registered ids remain real stories. A PR workflow runs the gate on PRs to `main` without a skip guard. Docker pins bumped for Alpine index drift and a writable `/app/node_modules/.cache` is created for the Playwright image. Docs include a one‑liner to regenerate the interaction manifest.

- Rollout
  - Add a dedicated `interaction` story export with a `play` function and register it in `tests/storybook/interaction-stories.json`; do not attach `play` to existing stories.
  - Use explicit `fn()` spies for asserted handlers.
  - Run `make test-storybook`; update the visual manifest so interaction stories stay pixel‑exempt.

<sup>Written for commit 6bca5976a14b5f99eaeb5a1ee0497e423beb91c5. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/VilnaCRM-Org/ui-toolkit/pull/133?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by RudoiDmytro** [comment]
@coderabbitai review

File: scripts/ci/story-play-functions.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+/** One exported story and the two facts the interaction gate cares about. */
+export interface ScannedStory {
+  exportName: string;
+  hasPlay: boolean;
+  tags: string[];
+}
+
+function objectLiteralOf(declaration: ts.VariableDeclaration): ts.ObjectLiteralExpression | null {
+  const initializer: ts.Expression | undefined = declaration.initializer;
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/story-play-functions.ts]: Fixed in b36deec — \`objectLiteralOf\` now unwraps \`as\`, \`satisfies\` and parenthesised expressions before checking for the object literal, so \`export const X = { play } satisfies Story\` is scanned too. Covered by a new unit case.

File: scripts/ci/storybook-interaction-manifest.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    .map(entry => entry.name.slice(0, -PLAY_TEST_SUFFIX.length))
+    .sort((a, b) => a.localeCompare(b));
+}
+
+/**
+ * Compares two key sets, returning a human-readable drift report or `null` when
+ * they are identical.
+ */
+export function formatDrift(expected: string[], actual: string[]): string | null {
+  const expectedSet: Set<string> = new Set(expected);
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/storybook-interaction-manifest.ts:131]: Fixed in b36deec — the gate now rejects a manifest that lists the same story twice (before the set-based drift comparison), and the unit drift guard asserts the committed registry has no duplicate rows.

File: src/components/ui-select-with-search/select-with-search.stories.tsx [hunk 3]
```diff
@@ -36,8 +57,24 @@ type Story = StoryObj<typeof UiSelectWithSearch>;
...
-    label: t('City'),
-    placeholder: t('Search city'),
+    label: cityLabel,
+    placeholder: cityPlaceholder,
     error: false,
   },
 };
+
+// Interaction story (`interaction` tag): proves the search text narrows the
+// listbox and the picked option becomes the field's value.
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply src/components/ui-select-with-search/select-with-search.stories.tsx:68]: Fair — fixed in b36deec by making the test earn the claim rather than softening it: both \`UiSelectWithSearch\` and \`UiSearchInput\` now assert a non-matching option (\`Odesa\` / \`Top performers\`) is absent from the listbox before clicking the match. Verified against the running Storybook.

File: scripts/ci/storybook-interaction-manifest.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    .sort((a, b) => a.id.localeCompare(b.id));
+}
+
+/** Stable comparison keys for a set of interaction stories. */
+export function interactionStoryKeys(stories: InteractionStory[]): string[] {
+  return stories.map(story => `${story.id} (${story.title} / ${story.exportName})`);
+}
+
+const TEST_CASE = /<testcase\b([^>]*?)(?:\/>|>([\s\S]*?)<\/testcase>)/g;
+const CASE_NAME = /\bname="([^"]*)"/;
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/storybook-interaction-manifest.ts:94]: Fixed in b36deec — the extracted \`name\` attribute is XML-decoded (\`&amp;\`, \`&lt;\`, \`&gt;\`, \`&quot;\`, \`&apos;\`, \`&#39;\`) before the suffix is stripped and compared, so an escaped title can no longer fail a passing run. Covered by a new unit case.

File: scripts/ci/story-play-functions.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+export function scanMetaTitle(fileName: string, source: string): string | null {
+  const sourceFile: ts.SourceFile = ts.createSourceFile(
+    fileName,
+    source,
+    ts.ScriptTarget.Latest,
+    true,
+    ts.ScriptKind.TSX
+  );
+
+  return sourceFile.statements.flatMap(titleOf)[0] ?? null;
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/story-play-functions.ts]: Fixed in b36deec — \`scanMetaTitle\` now resolves a direct \`export default { title: … }\` first and only falls back to the declared-object heuristic. Note this was never silent: the "reads a meta title out of every story file" test already failed loudly for an unreadable file; it is now simply supported.

**Comment by RudoiDmytro** [comment]
@coderabbitai review

File: scripts/ci/storybook-interaction-manifest.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/storybook-interaction-manifest.ts:131]: Hardened further in ba8ea06.

The earlier fix rejected duplicates only in the **id** key space (`id (title / exportName)`), but the JUnit reconciliation keys on `title exportName`. Two rows with *distinct ids* can still collide there, and a live index can legitimately expose both — so a single passing play test satisfied both rows. That is the same bypass you described, reached through the other key space.

The gate now rejects duplicates in **both** key spaces before either set comparison, and the success line reports the play tests actually reconciled rather than the rows registered, so the count can no longer overstate the evidence.

Verified against a crafted 14-row manifest whose extra row duplicates row 0's `title`+`exportName` under a new id, served by a matching live `index.json`:

```
storybook interaction gate: the interaction manifest lists the same story more than once: UiComponents/UiBackToMain KeyboardFocusReachesBackLink
```

The unit drift guard asserts the committed registry is duplicate-free in both key spaces too.

File: scripts/ci/run-storybook-interactions.ts [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+/**
+ * Resolves the caller-supplied manifest path against the project root and refuses
+ * anything that escapes it — the gate must never read (and then trust) a file from
+ * outside the repository. The containment check sits in the same function as the
+ * filesystem read so the sanitized path is the only one that can reach it.
+ */
+function readManifest(candidate: string): InteractionStory[] {
+  const resolved: string = resolve(PROJECT_ROOT, candidate);
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/run-storybook-interactions.ts:65]: Fixed in ba8ea06.

`readManifest` now checks containment **twice**: lexically first, so no `fs` call ever sees an unvalidated path, and then again on the `realpathSync` result — which is what closes the case you described. Both checks stay inside the reading function (32594df moved the guard there so SonarCloud's taint analysis sees the sanitizer), and `PROJECT_ROOT` stays a plain `resolve(process.cwd())` so nothing can throw at import time, outside `main`'s handler.

Verified by running the real script with `cwd` pointed at a throwaway root:

| argv | result |
| --- | --- |
| in-repo symlink → external file | refused |
| in-repo symlink → `/etc`, used as a path component | refused |
| `../outside/evil.json` | refused |
| absolute external path | refused |
| non-existent path *outside* the root | refused — the lexical check runs first, so this still reports the containment refusal instead of a missing-file error |
| non-existent path *inside* the root | missing-or-invalid-JSON |
| `tests/storybook/interaction-stories.json` | accepted, 13 rows |

Note that `realpathSync` collapses the whole chain, so a symlinked *intermediate* component is covered, not just the final one. The realpath→read window is left open deliberately and is now documented in the docblock: exploiting it needs write access to the checkout during the run, which already subsumes the attack.

File: src/components/ui-search-input/search-input.stories.tsx [hunk 7]
```diff
@@ -34,3 +55,23 @@ export const SearchInput: Story = {
...
+
+    await userEvent.type(field, typedQuery);
+    const match: HTMLElement = await screen.findByRole('option', { name: chosenSuggestion });
+
+    await expect(screen.queryByRole('option', { name: filteredOutSuggestion })).toBeNull();
+
+    await userEvent.click(match);
+
+    await expect(field).toHaveValue(chosenSuggestion);
+  },
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply src/components/ui-search-input/search-input.stories.tsx:72]: Fixed in ba8ea06. The duplicated block was the shared `play` body of the two typeahead fields. It now lives once in `.storybook/field-typeahead-interactions.ts` as `expectTypeaheadNarrowsAndSelects`, and both stories call it with their own labels and fixtures.

`qlty smells --all` no longer reports either story file. `make test-storybook` still passes 13/13 play tests in a real browser.

File: src/components/ui-search-input/search-input.stories.tsx [hunk 8]
```diff
@@ -34,3 +55,23 @@ export const SearchInput: Story = {
...
+export const SuggestionPickFillsField: Story = {
+  tags: ['interaction', '!autodocs'],
+  render: SearchInputInteractionStory,
+  play: async ({ canvasElement }): Promise<void> => {
+    const field: HTMLElement = within(canvasElement).getByRole('combobox', { name: searchLabel });
+
+    await userEvent.type(field, typedQuery);
+    const match: HTMLElement = await screen.findByRole('option', { name: chosenSuggestion });
+
+    await expect(screen.queryByRole('option', { name: filteredOutSuggestion })).toBeNull();
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply src/components/ui-search-input/search-input.stories.tsx]: Agreed and done in ba8ea06. The type → assert-filtered-out → click → assert-value sequence now lives once in `.storybook/field-typeahead-interactions.ts`; both stories pass their field name, query, chosen option and filtered-out option to it.

I named it `expectTypeaheadNarrowsAndSelects` rather than `expectFilterNarrowsAndSelects` because both callers are comboboxes and the helper is scoped to that widget, but the shape is exactly what you suggested. It sits in `.storybook/` next to `field-story-arg-types.ts` — that is the only place it can live, since `dependency-cruiser`'s `not-to-dev-dep` rule only exempts `*.stories.*` paths under `src/` from importing `storybook/test`.

This also cleared the two `qlty` similar-code findings on the same lines. Verified: `qlty smells --all` clean on both files, and the interaction suite still passes 13/13 in a real browser.

File: scripts/ci/story-play-functions.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    true,
+    ts.ScriptKind.TSX
+  );
+  const exported: string[] = sourceFile.statements.flatMap(defaultExportTitle);
+
+  if (exported.length > 0) {
+    return exported[0];
+  }
+
+  return sourceFile.statements.flatMap(declaredTitle)[0] ?? null;
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/story-play-functions.ts]: Valid — fixed in ba8ea06.

`scanMetaTitle` now resolves the expression `export default` actually names: an inline object literal is used directly, and an identifier is resolved back through its own top-level declaration. The "first declared `title`" fallback is gone, so a titled helper object above `meta` can no longer win, and a file whose default export cannot be resolved to an object now returns `null` and fails the "reads a meta title out of every story file" guard loudly instead of silently keying on the wrong component.

Regression cases added, including the one you asked for:

- `reads the title off the object `export default` names, not the first one` — a `{ title: 'Decoy/NotTheMeta' }` object declared *above* `meta`; the scan returns `Real/Meta`.
- `reports no title when the default export is missing or is not an object` — no default export, a call expression, and an undeclared identifier.

I also checked for behavioural drift on real data: both the committed and the new implementation return identical titles for all 26 `src/**/*.stories.tsx` files.

File: tests/unit/storybook-interaction-coverage.test.ts [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+const manifest: InteractionStory[] = JSON.parse(fs.readFileSync(manifestPath, 'utf8'));
+const allStories: StoryEntry[] = JSON.parse(fs.readFileSync(storiesManifestPath, 'utf8'));
+
+function storyFiles(directory: string): string[] {
+  return fs.readdirSync(directory, { withFileTypes: true }).flatMap(entry => {
+    const full: string = path.join(directory, entry.name);
+    if (entry.isDirectory()) return storyFiles(full);
+    return entry.name.endsWith('.stories.tsx') ? [full] : [];
+  });
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply tests/unit/storybook-interaction-coverage.test.ts:44]: Valid — fixed in ba8ea06, and widened past the two extensions you named.

`.storybook/main.ts` loads `*.stories.@(js|jsx|mjs|ts|tsx)`, so matching only `ts`/`tsx` would have left the same hole for a `.stories.js` module. Discovery now goes through `isStoryFile`, which mirrors that glob, and the scanner picks its `ts.ScriptKind` per extension — a `.stories.ts` parsed under TSX rules mis-reads a bare generic (`<T>(v: T) => v`) as an unclosed JSX tag and swallows the rest of the file. Measured on the new fixture: TSX parsing yields 1 of 4 top-level statements, dropping both `meta` and the story.

On the fixture: I did **not** add a real `src/**/*.stories.ts`, because a story file under `src/` has to be registered in `tests/visual/stories.json` and the interaction manifest, and would ship in the published library. Instead there are two fixtures with no such blast radius:

- a source-string `.stories.ts` module (bare generic + `meta` + a tagged `play` story) exercising `scanMetaTitle` and `scanStories`;
- `walks a real directory tree and picks up every story flavour in it` — an on-disk `mkdtempSync` tree containing `a.stories.ts`, `b.stories.tsx`, `c.stories.mdx`, `d.ts` and `nested/e.stories.js`, asserting the walk returns exactly the three story modules.

The same two-extension form was applied to the three configs that would otherwise have made `.stories.ts` support nominal only — `jest.config.ts` (coverage exclusion; the 100% gate would have failed on the first such file), `tsconfig.dts.json` (declaration build), and `.dependency-cruiser.js` (`no-orphans` allowlist and `no-prod-import-of-stories`) — all matching the form their sibling entries already use.

File: src/components/ui-select-with-search/select-with-search.stories.tsx [hunk 11]
```diff
@@ -36,8 +58,28 @@ type Story = StoryObj<typeof UiSelectWithSearch>;
...
+
+    await userEvent.type(field, typedQuery);
+    const match: HTMLElement = await screen.findByRole('option', { name: chosenCity });
+
+    await expect(screen.queryByRole('option', { name: filteredOutCity })).toBeNull();
+
+    await userEvent.click(match);
+
+    await expect(field).toHaveValue(chosenCity);
+  },
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply src/components/ui-select-with-search/select-with-search.stories.tsx:80]: Fixed in ba8ea06 — same root cause as the finding on `search-input.stories.tsx`. The shared `play` body now lives once in `.storybook/field-typeahead-interactions.ts` as `expectTypeaheadNarrowsAndSelects`, and both stories call it.

`qlty smells --all` no longer reports either story file.

File: scripts/ci/story-play-functions.ts [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+/**
+ * The object Storybook receives as the module's default export — either written
+ * inline (`export default { title: … }`) or named (`const meta = …; export
+ * default meta;`). A named export is resolved back to ITS declaration: reading
+ * the first titled object in the file instead would let an unrelated helper
+ * object declared above `meta` supply the wrong title.
+ */
+function defaultExport(sourceFile: ts.SourceFile): ts.Expression | undefined {
+  const assignment: ts.ExportAssignment | undefined = sourceFile.statements.find(
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/story-play-functions.ts:167]: Valid — fixed in 6bca597.

`defaultExport` only looked at `ExportAssignment` nodes, so `export { meta as default }` fell through to `undefined` and the drift guard rejected the file as untitled. It now falls back to scanning `ExportDeclaration` nodes for a specifier named `default` and resolves its `propertyName` through the same local-declaration lookup, so the alias form behaves exactly like `export default meta`.

Re-exports are deliberately excluded: `export { meta as default } from './other-meta'` carries a module specifier, so there is no local declaration to read and the scan still returns `null` rather than guessing.

Two regression cases added — `resolves a meta aliased with \`export { meta as default }\`` (with a decoy titled object declared above `meta`, so it also pins the resolution order) and `does not resolve a default re-exported from another module`.

File: scripts/ci/run-storybook-interactions.ts [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+  try {
+    canonical = realpathSync(resolved);
+  } catch {
+    return fail(UNREADABLE);
+  }
+
+  const canonicalToRoot: string = relative(realpathSync(PROJECT_ROOT), canonical);
+
+  if (canonicalToRoot === '' || canonicalToRoot.startsWith('..') || isAbsolute(canonicalToRoot)) {
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/run-storybook-interactions.ts]: Valid — fixed in 6bca597, at both call sites (the lexical pre-check and the canonical re-check).

I used `path.sep` rather than hardcoding both separators, since `path.relative` returns platform-native separators and a POSIX path segment cannot contain `/`:

```ts
const PARENT_SEGMENT: string = `..${sep}`;
…
relativeToRoot === '' ||
relativeToRoot === '..' ||
relativeToRoot.startsWith(PARENT_SEGMENT) ||
isAbsolute(relativeToRoot)
```

Verified against a throwaway root containing a real `..dotdir` directory:

| argv | before | after |
| --- | --- | --- |
| `..dotdir/m.json` (in-root) | refused | **accepted** |
| symlink → in-root `..dotdir` | refused | **accepted** |
| symlink → outside the root | refused | refused |
| `../outside/evil.json` | refused | refused |
| `tests/storybook/interaction-stories.json` | accepted | accepted |

File: scripts/ci/run-storybook-interactions.ts [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+async function main(): Promise<void> {
+  const url: string = trimTrailingSlashes(process.env.REACT_APP_STORYBOOK_URL ?? DEFAULT_URL);
+  const manifest: InteractionStory[] = readManifest(process.argv[2] ?? DEFAULT_MANIFEST);
+
+  assertManifestMatchesIndex(manifest, await fetchLiveStories(url));
+  runTestRunner(url);
+  // Report what was reconciled, not what was registered, so the count is evidence.
+  const passed: string[] = assertEveryStoryRan(manifest);
+
+  process.stdout.write(`storybook interaction gate: ${passed.length} play tests passed.\n`);
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/run-storybook-interactions.ts:228]: Valid — fixed in 6bca597. `assertEveryStoryRan` now returns `[...new Set(passed)]`, so the printed count is distinct reconciled stories rather than raw JUnit rows, with the reasoning recorded at the return site.

Worth noting the drift check itself was already safe here — `formatDrift` set-compares, so a repeated `<testcase>` row cannot mask a missing story. It was only the printed number that could overstate, which is exactly what you flagged, and the count is meant to read as evidence.

File: .dependency-cruiser.js [hunk 15]
```diff
@@ -46,7 +46,7 @@ module.exports = {
           '\\.d\\.ts$', // type declarations (Types.d.ts, styles.d.ts, react-app-env.d.ts)
           '(^|/)tsconfig\\.json$',
           '(^|/)(?:babel|webpack)\\.config\\.(?:js|cjs|mjs|ts|json)$',
-          '\\.stories\\.tsx$', // Storybook stories are entry points, not orphans
+          '[.]stories[.](?:tsx?|jsx?)$', // Storybook stories are entry points, not orphans
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply .dependency-cruiser.js]: Valid — fixed in 6bca597, and extended past the two patterns you named.

`.mjs` is now in both the `no-orphans` allowlist and `no-prod-import-of-stories`. While checking, the same gap existed in two more rules that share the pattern — `peer-deps-used` and `not-to-dev-dep` (both `[.](?:spec|test|stories)[.](?:tsx?|jsx?)$`), where a `.stories.mjs` importing `storybook/test` would have failed the dev-dependency gate. All four carry `mjs` now.

The four pattern locations are back in sync with `.storybook/main.ts`'s `@(js|jsx|mjs|ts|tsx)`:

| location | pattern |
| --- | --- |
| `.storybook/main.ts` | `*.stories.@(js\|jsx\|mjs\|ts\|tsx)` |
| `scripts/ci/story-play-functions.ts` | `/\.stories\.(?:mjs\|jsx?\|tsx?)$/` |
| `jest.config.ts` | `!src/**/*.stories.{js,jsx,mjs,ts,tsx}` |
| `tsconfig.dts.json` | one exclude per flavour |
| `.dependency-cruiser.js` | `[.]stories[.](?:mjs\|tsx?\|jsx?)$` ×4 |

`depcruise` still reports 0 errors and the declaration build is clean.

**Comment by RudoiDmytro** [comment]
@coderabbitai approve

**Comment by RudoiDmytro** [comment]
@coderabbitai approve

**Comment by RudoiDmytro** [comment]
@coderabbitai approve

**Comment by RudoiDmytro** [comment]
@coderabbitai review

File: scripts/ci/run-storybook-interactions.ts [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+function trimTrailingSlashes(url: string): string {
+  let end: number = url.length;
+  while (end > 0 && url.charAt(end - 1) === '/') {
+    end -= 1;
+  }
+  return url.slice(0, end);
+}
+
+async function fetchLiveStories(url: string): Promise<InteractionStory[]> {
+  const response: Response = await fetch(`${url}/index.json`);
```
[the remark this answers is not in the document]
Comment (RudoiDmytro) [reply scripts/ci/run-storybook-interactions.ts]: Fixed in 8b953fb.

`fetchLiveStories` now wraps the request in an `AbortController` with a 60s budget and clears the timer in a `finally`, so it is cleared only after `readIndex` has finished reading the body — not before. The status/body handling moved into its own `readIndex` function so the abort path stays legible, and an aborted request is reported as `the Storybook index did not answer within 60000ms` rather than surfacing a bare `AbortError`.

Probed against a server that accepts the connection and never responds (temporarily shortening the constant to 1500ms): the gate exits 1 after ~1.5s with that message instead of hanging to the job timeout. The other paths are unchanged — a 503 still reports `the Storybook index request failed with status 503`, an empty index still reports manifest drift, and the real index still reconciles and proceeds to the runner.

**Comment by RudoiDmytro** [comment]
@coderabbitai review

**Review by Kravalg — APPROVED** [review]
